### PR TITLE
Yes hi this does the thing.

### PR DESCRIPTION
**Pull request by Zundrel**
Welcome to my Thing storage API, it let's you store things in a perfectly normal and efficient manner (and totally not a horrible one). It functions like the universe, literally.

You start with a Thing. The Thing can be stored in a Capsule which can be stored in a Box which can be stored in a Crate which can be stored in a Warehouse which can be stored in a Planet which can be stored in a Dimension which can be stored in a Universe.

is good. feel free to do the thing to my thing to make it even more thing-like if you like

oh and also because of the randomness of the universe, warehouses and above can only store a random amount of things between 20 and 40 things, including the universe itself

i hope this limitation makes sense to you, as randomness has to be emulated for a storage api to be a good thing, at least i think. thats what some random guy on reddit told me.

### CODE REVIEW

**Comment by vini2003** [comment]
Missing `IThing`, `IBox`, `ITetrahedron` and a few others.

**Comment by Zundrel** [comment]
ah wow, how could i forget these important storage api things, i will add the things to the thing now

File: src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package io.github.fabriccommunity.everything.storage;
+
+// Stores a thing, but it's abstract now.
+public abstract class AbstractThing<T> {
+	T thing;
+
+	// Don't do this, very bad.
+	public AbstractThing() {
+		thing = (T) "wababa";
```
Comment (liach) [inline src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java:9]:
```suggestion
		this((T) "wababa");
```

File: src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package io.github.fabriccommunity.everything.storage;
+
+// Stores a thing, but it's abstract now.
+public abstract class AbstractThing<T> {
```
Comment (i509VCB) [inline src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java:4]: Should be `ThingAbstract`

Comment (liach) [reply src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java:4]: Should be `Abstract`

Comment (vini2003) [reply src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java:4]: Should be `feb5f4784065401e896fe5e8055125f5`

**Comment by Zundrel** [comment]
IBox, ITetrahedron, and ThingHandler are now a thing.

File: src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java (hunk 2 above)
Comment (Zundrel) [reply src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java]: did the thing

File: src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java (hunk 1 above)
Comment (Zundrel) [reply src/main/java/io/github/fabriccommunity/everything/storage/AbstractThing.java]: did the thing

**Comment by Zundrel** [comment]
ready for merge unless anyone can make it more ~~awful~~ great

**Comment by vini2003** [comment]
we're lacking builders, builder factories and builder factory builders

**Comment by Devan-Kerman** [comment]
ThingBeanFactoryBuilderContainerConsumerOpticalMonadFunctor please

File: src/main/java/io/github/fabriccommunity/everything/storage/Warehouse.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+		thingy.forTheThingsPerformTheThing();
+		return "Yes hi, thing in the thing now.";
+	}
+
+	public String addThing(Crate thingy) {
+		things.add(thingy);
+		thingy.forTheThingsPerformTheThing();
+		return "Um thingy in the doohickey now";
+	}
+}
```
Comment (valoeghese) [inline src/main/java/io/github/fabriccommunity/everything/storage/Warehouse.java:35]: In my 10 seconds scanning code it seems fine. merg

**Pull request merged**